File: Lv.1/ex32.js
//프로그래머스 정수 제곱근 판별
function solution(n) {
    var answer = 0;
    //양의 정수 n이 어떤 정수의 제곱이라면, Math.sqrt(n)의 정수부분만 추출한결과가 Math.sqrt(n)이랑 같을것이다.
    if(parseInt(Math.sqrt(n)) === Math.sqrt(n))
        return (Math.sqrt(n)+1)*(Math.sqrt(n)+1);
    else
        return -1;
}

//주어진 정수가 어떤 정수의 제곱인지 판별하는 조건을 구하는게 핵심이었다고 생각한다, n이 어떤 정수의 제곱이 아닌 경우를
//어떻게 판별해야하는지 고민을 했고, 어떤 정수의 제곱이 아니라면 parseInt()로 Math.sqrt(n)의 정수부분만 추출한 값과 Math.sqrt(n)의 값이
//다르다는걸 조건으로 삼아 구분을 했다.

//다른 코드들을 구경해보면서 어떤 수의 제곱을 구하고싶을때 두번쓰는것이아니라 Math.pow(a,b) = a**b 라는것을 알게되었다.
//괜찮은 풀이가 있다, Number.isInteger()를 이용했는데, 검색해보니 isInteger()는 전달된 값이 정수인지 여부를 bool값으로 리턴시켜준다고 한다.
//이거를 이용하면 내가 조건문에 적었던 조건식을 대체가 가능하다.
function nextSqaure(n){
    var result = 0;
    var n = Math.sqrt(n);
    result = Number.isInteger(n) ? Math.pow(n+1, 2) : 'no';
    return result;
}

File: Lv.1/ex32.java
//시저 암호

//아스키코드 값을 이용하는 과정에서 미리 알파벳정도는 아스키코드 값을 알필요가 있을것같다.
// A = 65, a = 97 두개만 기억하자 나머지는 직접 따져볼수 있다.
public class ex32 {
    //A 65 a 97 z보다 ascii 값이 크면 더한 값에서 -26
    public String solution(String s, int n) {
        String answer = "";
        
        char[] sArr = s.toCharArray();
        for(int i=0; i<sArr.length; i++){
            if(sArr[i] == ' '){
                answer += sArr[i];
                continue;
            }
            if(sArr[i]>= 'a' && sArr[i]<='z'){
                if(sArr[i]+n >'z'){
                    answer += (char)(sArr[i]+n-26);
                }else{
                    answer += (char)(sArr[i]+n);
                }
            }else if(sArr[i]>='A' && sArr[i] <= 'Z'){
                if(sArr[i]+n >'Z'){
                    answer += (char)(sArr[i]+n-26);
                }else{
                    answer += (char)(sArr[i]+n);
                }
            }
        }
        
        return answer;
    }
}


//다른 풀이
//char 타입의 대소문자 판별기능이 있다. 
//Character.isLowerCase(), .isUpperCase()
class Caesar {
    String caesar(String s, int n) {
        String result = "";
    n = n % 26;
    for (int i = 0; i < s.length(); i++) {
      char ch = s.charAt(i);
      if (Character.isLowerCase(ch)) {
        ch = (char) ((ch - 'a' + n) % 26 + 'a');
      } else if (Character.isUpperCase(ch)) {
        ch = (char) ((ch - 'A' + n) % 26 + 'A');
      }
      result += ch;
    }
        return result;
    }

    public static void main(String[] args) {
        Caesar c = new Caesar();
        System.out.println("s는 'a B z', n은 4인 경우: " + c.caesar("a B z", 4));
    }
}